File: training-6.0/6-3i.py
# https://contest.yandex.ru/contest/66794/problems/I/
from collections import deque

reader = open('input.txt', 'r')
n = int(reader.readline())
a, b = map(int, reader.readline().split())
if b < a and not (a == 4 and b == 1) or (a == 1 and b == 4):
    a, b = b, a
order = [a - 1, b - 1]
if b - a == 2:
    is_direct = True
    order += [a % 4, b % 4]
else:
    is_direct = False
    order += [b % 4, (b % 4 + 1) % 4]
rovers = [0] * n
for i in range(n):
    d, t = map(int, reader.readline().split())
    rovers[i] = (t, order.index(d - 1), i)
reader.close()
rovers.sort()
cur_time = rovers[0][0]
cnt, res, it = 0, [0] * n, [0] * 4
for i in range(4):
    rovers.append((1000, i, 101))
    while rovers[it[i]][1] != i: it[i] += 1
while cnt < n:
    a = 0
    for j in range(4):
        if j == 2 and a != 0: break
        if rovers[it[j]][0] <= cur_time:
            res[rovers[it[j]][2]] = cur_time
            it[j] += 1
            while rovers[it[j]][1] != j: it[j] += 1
            n -= 1
            if not is_direct: break
            a += 1
    cur_time += 1
print(*res, sep='\n')

File: training-6.0/6-3i.cpp
// https://contest.yandex.ru/contest/66794/problems/I/
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <deque>
using namespace std;

int main() {
	ifstream cin("input.txt");
	size_t n, a, b;
	cin >> n >> a >> b;
	vector<size_t> order (5);
	bool isDirect;
	if (b < a and not (a == 4 and b == 1) or (a == 1 and b == 4))
		swap(a, b);
	order[a] = 0; 
	order[b] = 1;
	if (b - a == 2) {
		isDirect = true;
		order[a % 4 + 1] = 2;
		order[b % 4 + 1] = 3;
	}
	else {
		isDirect = false;
		order[b % 4 + 1] = 2;
		order[(b % 4 + 1) % 4 + 1] = 3;
	}
	vector<tuple<size_t, size_t, size_t>> rovers(n);
	size_t d, t;
	for (size_t i = 0; i < n; i++) {
		cin >> d >> t;
		rovers[i] = { t, order[d], i };
	}
	for (size_t i = 0; i < 4; i++)
		rovers.push_back({ 1000001, i, 1000 });
	sort(rovers.begin(), rovers.end());
	size_t cnt = 0, cur_time = get<0>(rovers[0]);
	vector<size_t> res(n);
	deque < pair<size_t, size_t>> dq[4];
	for (auto [t, o, i] : rovers) 
		dq[o].push_back({ t,i });
	while (cnt < n) {
		bool isFree = true, canTry = false;
		for (size_t j = 0; j < 4; j++) {
			if (j % 2 == 0)
				canTry = isFree;
			else
				canTry = isDirect or isFree;
			if (!canTry) break;
			auto [t, i] = dq[j].front();
			if (t <= cur_time) {
				res[i] = cur_time;
				dq[j].pop_front();
				cnt++;
				isFree = false;
			}
		}
		cur_time++;
	}
	for (auto e : res)
		cout << e << '\n';
	return 0;
}
